### ISSUE TITLE
Periodically evaluating on a validation set

### ISSUE THREAD

**Issue by varunkumar3618** [tensorflow/tensorflow#7902]
A common training pattern is to run an epoch, or a few epochs, of training and then evaluate on a development set. As far as I can tell, there does not seem to be an easy way to do this while also using the TFRecords format. While it is possible to run through a data split by setting the num_epochs parameter in the functions to read in single examples, doing so would require the computation graph to be rebuilt before every validation phase. Instead, it would be preferable to have a symbolic example that could be reset to the beginning of the dataset using a related operation.

I feel that the above pattern is sufficiently widely used to make a new feature worthwhile. Further, having control over the number of epochs also becomes important when trying to replicate the performance numbers of other implementations.

####
The same training-validation pattern using feed dicts:
for epoch in num_epochs:
  for batch in get_batches(train, epoch):
    sess.run(train_op, feed_dict={"batch": batch)
  num_correct, total = 0, 0
  for batch in get_batches(val, epoch):
    acc = sess.run(acc_op, feed_dict={"batch", batch)
    total += batch.shape[0]
    num_correct += batch.shape[0] * acc
  print "Dev accuracy after %s epochs: %s" % (epoch + 1, num_correct / total)

**Comment by vonclites** [comment]
Have you looked at [Validation Monitors](https://www.tensorflow.org/get_started/monitors#configuring_a_validationmonitor_for_streaming_evaluation)? I have not tried to use this yet. I'm not sure how much it will require you to rework your code to fit the tf.learn format. My understand is that the validation monitor would take an input function. That input function would provide data via whatever method you want to use, including reading TFRecord examples.

**Comment by varunkumar3618** [comment]
Thanks for the suggestion. However, I don't think switching to tf.learn would really solve the issue. The hooks allow you to run a validation phase after a certain number of steps, which I can do well enough in my own code, but still rely on underlying queues to determine when the data is over. As a result, the limitation in the io utilities applies equally to them. (i.e. the lack of a method to read an epoch of data multiple times without rebuilding the graph).

**Comment by inflation** [comment]
In my understanding, if you use queues as the input source, it's kind of "built in" to the computation graph. So I would like to use a `tf.cond` which read a boolean and determine which function to be used. Then, you can change the value of the boolean on the go.

**Comment by varunkumar3618** [comment]
That makes sense, I'll try it out.

**Comment by tatatodd** [comment]
@varunkumar3618 

Please check out #7951 , which acknowledges the deficiency here and proposes an enhancement moving forwards.  Please add any use-cases onto that issue, so we can collect them in one place!

Since I believe this issue was intended as a feature request, I'm closing it out so that we can focus on the other one.  If I've made a mistake in this regard, feel free to ping this issue and I'll re-open it.  Thanks!

**Comment by varunkumar3618** [comment]
Yes, thank you. That's exactly what I was requesting.

**Comment by wenmin-wu** [comment]
have you found any adhoc way to solve this problem? If you have, please let me know. thx!

**Comment by vonclites** [comment]
The new tf.data API should suit your needs. https://www.tensorflow.org/api_docs/python/tf/data
